designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,271 @@
+# Disruption Controls By Reason 
+## User Scenarios 
+1. Users need the capability to schedule upgrades only during business hours or within more restricted time windows. Additionally, they require a system that doesn't compromise the cost savings from consolidation when upgrades are blocked due to drift.
+2. High-Frequency Trading (HFT) firms require full compute capacity during specific operational hours, making it imperative to avoid scale-down requests for consolidation during these periods. However, outside these hours, scale-downs are acceptable.
+See Less Made Up Scenarios here: 
+- https://github.com/kubernetes-sigs/karpenter/issues/924 
+- https://github.com/kubernetes-sigs/karpenter/issues/753#issuecomment-1790110838
+- https://github.com/kubernetes-sigs/karpenter/issues/672
+
+## Known Requirements 
+**Reason and Budget Definition:** Users should be able to define an reason and a corresponding budget(s).
+**Supported Reasons:** All disruption Reasons affected by the current Budgets implementation (Consolidation, Emptiness, Expiration, Drift) should be supported. 
+**Default Behavior for Unspecified Reasons:** Budgets should continue to support a default behavior for all disruption reasons. If an reason is unspecified, it is assumed to apply to all reasons. If a reason is unspecified, we apply the budget to be shared by all reasons that are unspecified. 
+
+## API Design
+### Approach A: Add a reason field to disruption Budgets 
+This approach outlines a simple api change to the v1beta1 nodepool api to allow disruption budgets to specify a list of disruption methods. 
+
+### Proposed Spec
+Add a simple field "reasons" is proposed to be added to the budgets. 
+```go
+// Budget defines when Karpenter will restrict the
+// number of Node Claims that can be terminating simultaneously.
+type Budget struct {
+      // Reasons is a list of methods for disruption that apply to this budget. If Reasons is not set, this budget applies to all methods.
+      // If a reason == "default", it will apply to all reasons that don't have an active budget. If a reason is set, it will only apply to that method. If multiple reasons are specified,
+      // this budget will apply to all of them. If a budget is not specified for a method, the default budget will be used.
+      // allowed reasons are "default", "consolidation", "expiration", "emptiness", "drift"
+      // +kubebuilder:validation:MaxItems=5
+      // +kubebuilder:validation:Enum:={"default","consolidation","expiration","emptiness","drift"}
+      // +optional
```

Comment:
#### Q: Should default be specifed only as omitted? Or should users be able to define default? 
We talked about having "default" as an action case. Meaning that if I specify a budget for consolidation and Drift, then a separate budget with an empty reason in the same active window will be taken as the default number of disruptable nodes in a given window for all actions that were not defined. Should we also be able to define default as a reason explicitly? 

The answer is yes to reduce redundancy. 

I can define a single budget to cover all reasons for disruption like so 

```
  - nodes: 10
  reasons: [Drift, Consolidation, Default]
  schedule: "* * * * *"
```
Rather than having to specify two budgets like 
```
  - nodes: 10
  reasons: [Drift, Consolidation]
  schedule: "* * * * *"
  - nodes: 10
  schedule: "* * * * *"
```

To communicate the same thing to karpenter. So we should allow users to specify default directly and explicitly, alongside having the behavior for default from specifying like 
```
  - nodes: 10
  schedule: "* * * * *"
```